File: LeetCode/Problems/Python/251. Flatten 2D Vector.py
#Time: O(v/n), n is the total amount of integer, v is vectors, the total is (n+v)/n 
#Space: O(1)
class Vector2D:

    def __init__(self, v: List[List[int]]):
        self.vector = v
        self.x = 0
        self.y = 0
        

    def next(self) -> int:
        while self.x < len(self.vector) and self.y == len(self.vector[self.x]):
            self.x+=1
            self.y=0
        ret = self.vector[self.x][self.y]
        self.y+=1
        return ret
        

    def hasNext(self) -> bool:
        while self.x < len(self.vector) and self.y == len(self.vector[self.x]):
            self.x+=1
            self.y=0
        return self.x < len(self.vector)


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(v)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: LeetCode/Problems/C++/251. Flatten 2D Vector.cpp
class Vector2D {
public:
    Vector2D(vector<vector<int>>& v) {
        my_vector = &v;
        x = 0;
        y = 0;
    }
    
    int next() {
        hasNext();
        return (*my_vector)[y][x++];
    }
    
    bool hasNext() {
        
        while(y<my_vector->size() && x==(*my_vector)[y].size()){
            x=0;
            y++;
        }
        return y<my_vector->size();
    }
private:
    int x, y;
    vector<vector<int>>* my_vector;
};


class Vector2D {
public:
    Vector2D(vector<vector<int>>& v) {
        it = v.begin();
        end = v.end();
        x = 0;
       
    }
    
    int next() {
        hasNext();
        return (*it)[x++];
    }
    
    bool hasNext() {
        
        while(it!=end && x==(*it).size()){
            x=0;
            it++;
        }
        return it!=end;
    }
private:
    int x;
    vector<vector<int>>::iterator it, end;
};
